src/interpreter/interpreter.kt

```diff
@@ -0,0 +1,195 @@
+        InOp.leq -> VBool(iLeft <= iRight)
+        InOp.geq -> VBool(iLeft >= iRight)
+        InOp.and -> VBool(bLeft && bRight)
+        InOp.or -> VBool(bLeft || bRight)
+        InOp.concat -> VString(sLeft + sRight)
+    }
+}
+
+fun <K, V> MutableMap<K, V>.copy() = this.toMutableMap()
+
+fun retPass(function : TommyFunc) : TommyFunc {
+    val (id, _, _, body, _) = function
+    val retVar = Var("$id+")
+    val retFlag = Var("$id&")
+    val newBody : MutableList<AST> =
+            mutableListOf(UntypedVarDef(retVar, LUnit), UntypedVarDef(retFlag, LBool(false)))
+    fun rewriteReturn(ast : AST) : AST {
+        fun wrap(curr : AST) = IfStep(Prefix(PreOp.not, retFlag), listOf(curr), null)
+        fun rewriteIf(ifS : If) : If = when(ifS) {
+            is IfStep -> IfStep(ifS.cond, ifS.body.map(::rewriteReturn), ifS.next?.let { rewriteIf(it) })
+            is Else -> Else(ifS.body.map(::rewriteReturn))
+        }
+        return when (ast) {
+            is Return -> {
+                IfStep(Prefix(PreOp.not, retFlag),
+                        listOf(VarReassign(retVar, ast.toReturn), VarReassign(retFlag, LBool(true))), null)
+            }
+            is IfStep -> wrap(rewriteIf(ast))
+            is Else -> wrap(Else(ast.body.map(::rewriteReturn)))
+            is While -> wrap(While(ast.cond, ast.body.map(::rewriteReturn)))
+            is For -> wrap(For(ast.elemIdent, ast.list, ast.body.map(::rewriteReturn)))
+            is Expr, is UntypedVarDef, is VarDef, is VarReassign, is ArrayAssignment, is FunDef -> wrap(ast)
+        }
+    }
+    newBody.addAll(body.map(::rewriteReturn))
+    return function.copy(statements = newBody)
+}
+
+fun eval(expr : Expr, environment : MutableMap<String, Value>,
+             functionDefs : MutableMap<String, Func>) : Value = when (expr) {
+    is LInt -> VInt(expr.value)
+    is LString -> VString(expr.value)
+    is LBool -> VBool(expr.value)
+    is LArray -> VArray(expr.value.mapTo(mutableListOf()) { eval(it, environment, functionDefs) })
+    is LUnit -> VUnit
+    is Prefix -> expr.op.eval(eval(expr.arg, environment, functionDefs))
+    is Infix -> expr.op.eval(eval(expr.lhs, environment, functionDefs), eval(expr.rhs, environment, functionDefs))
+    is Var -> environment[expr.id] ?: throw UndefinedVariableException(wrongAST = expr, wrongId = expr.id)
+    is FunCall -> {
+        val func = functionDefs[expr.id.id]
+                ?: throw UndefinedVariableException(wrongAST = expr, wrongId = expr.id.id)
+        if(func.args.size != expr.args.size) throw IncorrectArgumentCountException(wrongCall = expr, called = func)
+        val evaledArgs = expr.args.map { argExpr -> eval(argExpr, environment, functionDefs) }
+        when(func) {
+            is TommyFunc -> {
+                val (id, argAnnIds, _, body, functionEnv) = func
+                val argIds = argAnnIds.map { (argId, _) -> argId }
+                val localCtx = functionEnv.toMutableMap()
```

Comment:
functionEnv is the environment at time of funDef.
shouldn't the statements in a function be called in the environment of the caller, in this case `environment` (with args mixed in of course)

Reply:
That's dynamic scoping, and basically the only languages to ever use it are early variants of lisp (including emacs lisp).

What should the following program do?
```
let example() : Int = return y
let y : Int = 5
println(example())
```
In your proposed semantics it would successfully print `5`, but in mine it would error.

My implementation does have the issue that it freezes all the variable definitions, but I've been thinking about how to change that. My current idea is to change the environment type to `MutableSet<VarDef>`, where `data class VarDef(val x : String, var v : Value)`. Then when we update variables we mutate the `VarDef` entry in our set instead of assigning a new value to a certain key for a specific map.

Reply:
That example program is almost exactly what I checked in python when being confused by this part of your code, so apparently python also has dynamic scoping. ruby doesn't.

personally I agree that we should not have dynamic scoping.

however, our current implementation DOES have dynamic typing, according to my tests.
That is why I was extremely confused by the behavior/implementation of this. I guess it's a bug. My guess is that `functionEnv` and `environment` are actually pointing to the same variable.

Reply:
what do you mean by freezes all the variable definitions?

Reply:
We have dynamic typing because I've been too busy to write the typechecker. Also, python does not have dynamic scope, I just gave a bad example. Python seems to have `y` be a variable set to `None` or something, and then when the function is called it just gets the current value of `y` (like you'd expect). You can see it doesn't have dynamic scope because this program will print 5:
```
def f(): 
  x = 5 
  return lambda: x 

x = 7
print(f()())
```

Reply:
functionEnv and environment are definitely not pointing to the same variable, functionEnv is a shallow copy of environment

Reply:
By "freezes variable definitions" I mean that this program will print `5` in my semantics:
```
let x = 5
let f() : Unit = x
x = 7
println(f())
```

Reply:
ok. The python thing that confused me was a quirk with python and the global scope. Any level deeper and it works as expected. Your solution to the variable freezing sounds good